videoio_dynamic.basic_write sporadically fails on CI

### System Information

Platform: Ubuntu 20.04 ARM64
OpenCV version: 4.8.0-dev


### Detailed description

```
[ RUN      ] videoio_dynamic.basic_write

(opencv_test_videoio:14940): GStreamer-CRITICAL **: 13:37:41.417: gst_element_make_from_uri: assertion 'gst_uri_is_valid (uri)' failed
[ERROR:0@2.430] global cap.cpp:643 open VIDEOIO(CV_IMAGES): raised OpenCV exception:

OpenCV(4.8.0-dev) /home/ci/opencv/modules/videoio/src/cap_images.cpp:431: error: (-215:Assertion failed) !filename_pattern.empty() in function 'open'



(opencv_test_videoio:14940): GStreamer-CRITICAL **: 13:37:42.594: gst_query_set_position: assertion 'format == g_value_get_enum (gst_structure_id_get_value (s, GST_QUARK (FORMAT)))' failed
[ WARN:0@3.144] global cap_gstreamer.cpp:1728 open OpenCV | GStreamer warning: Cannot query video position: status=1, value=-1, duration=120
[ WARN:0@3.635] global cap.cpp:204 open VIDEOIO(V4L2): backend is generally available but can't be used to capture by name
/home/ci/opencv/modules/videoio/test/test_dynamic.cpp:73: Failure
Expected equality of these values:
  count
    Which is: 0
  FRAME_COUNT
    Which is: 120
Corrupt JPEG data: premature end of data segment
Corrupt JPEG data: premature end of data segment
Corrupt JPEG data: premature end of data segment
Corrupt JPEG data: premature end of data segment
Corrupt JPEG data: premature end of data segment
[  FAILED  ] videoio_dynamic.basic_write (3466 ms)
```

### Steps to reproduce

CI example: https://github.com/opencv/opencv/actions/runs/5464664040/jobs/9947005647?pr=23881

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

--- a/modules/videoio/src/cap_images.cpp
+++ b/modules/videoio/src/cap_images.cpp
@@ -302,59 +302,58 @@ bool CvCapture_Images::open(const std::string& _filename)
     if (filename_pattern.empty())
     {
         filename_pattern = _filename;
-        cv::String filename = _filename.c_str();
-        if (!utils::fs::exists(filename))
+        if (!utils::fs::exists(filename_pattern))
         {
+            CV_LOG_INFO(NULL, "CAP_IMAGES: File does not exist: " << filename_pattern);
+            close();
             return false;
         }
-        if (!haveImageReader(filename))
+        if (!haveImageReader(filename_pattern))
         {
-            CV_LOG_INFO(NULL, "CAP_IMAGES: Stop scanning. Can't read image file: " << filename);
+            CV_LOG_INFO(NULL, "CAP_IMAGES: File is not an image: " << filename_pattern);
+            close();
+            return false;
         }
         length = 1;
-        // grab frame to enable properties retrieval
-        bool grabRes = grabFrame();
-        grabbedInOpen = true;
-        currentframe = 0;
-        return grabRes;
     }
-
-    // determine the length of the sequence
-    for (length = 0; ;)
+    else
     {
-        cv::String filename = cv::format(filename_pattern.c_str(), (int)(offset + length));
-        if (!utils::fs::exists(filename))
+        // determine the length of the sequence
+        for (length = 0; ;)
         {
-            if (length == 0 && offset == 0) // allow starting with 0 or 1
+            cv::String filename = cv::format(filename_pattern.c_str(), (int)(offset + length));
+            if (!utils::fs::exists(filename))
             {
-                offset++;
-                continue;
+                if (length == 0 && offset == 0) // allow starting with 0 or 1
+                {
+                    offset++;
+                    continue;
+                }
+                CV_LOG_INFO(NULL, "CAP_IMAGES: File does not exist: " << filename);
+                break;
             }
-            break;
+
+            if(!haveImageReader(filename))
+            {
+                CV_LOG_INFO(NULL, "CAP_IMAGES: File is not an image: " << filename);
+                break;
+            }
+
+            length++;
         }
 
-        if(!haveImageReader(filename))
+        if (length == 0)
         {
-            CV_LOG_INFO(NULL, "CAP_IMAGES: Stop scanning. Can't read image file: " << filename);
-            break;
+            close();
+            return false;
         }
 
-        length++;
+        firstframe = offset;
     }
-
-    if (length == 0)
-    {
-        close();
-        return false;
-    }
-
-    firstframe = offset;
-
     // grab frame to enable properties retrieval
-    bool grabRes = grabFrame();
+    bool grabRes = CvCapture_Images::grabFrame();
     grabbedInOpen = true;
     currentframe = 0;
-
     return grabRes;
 }
 

